vision/medical_imaging/3d-unet-kits19/base_SUT.py

```diff
@@ -0,0 +1,195 @@
+    SUT works with QSL (Query Sample Library) to locate queries from LoadGen.
+    SUT works with backend (or framework such as PyTorch, ONNX-runtime, TensorFlow),
+    to perform inference of the given queries and reports back to LoadGen.
+
+    It is required to implement backend specific SUT; inheriting this baseline SUT
+    enables minimal implementation of the backend specific SUT.
+
+    Attributes
+    ----------
+    sut: object
+        SUT in the context of LoadGen
+    qsl: object
+        QSL in the context of LoadGen
+    model_path: str or PosixPath object
+        path to the model for backend
+
+    Methods
+    -------
+    to_tensor(my_array):
+        transforms my_array into tensor form backend understands
+    from_tensor(my_tensor):
+        transforms my_tensor backend worked on into numpy friendly array
+    do_infer(input_tensor):
+        performs backend specific inference upon input_tensor
+    infer_single_query(data, mystr):
+        performs inference upon data and summarize work in mystr
+    issue_queries(query_samples):
+        LoadGen calls this with query_samples, a vector containing series of queries
+        SUT is to perform each query and calls back to LoadGen with QuerySamplesComplete()
+    flush_queries():
+        not used
+    process_latencies(latencies_ns):
+        not used
+    """
+
+    def __init__(self, preprocessed_data_dir, performance_count):
+        """
+        Constructs all the necessary attributes for ONNX Runtime specific 3D UNet SUT
+        Baseline SUT doesn't instantiate any 3D UNet model; backend specific SUT needs
+        to properly instantiate model
+
+        Parameters
+        ----------
+            preprocessed_data_dir: str or PosixPath
+                path to directory containing preprocessed data
+            performance_count: int
+                number of query samples guaranteed to fit in memory
+        """
+        print("Constructing SUT...")
+        self.sut = lg.ConstructSUT(
+            self.issue_queries, self.flush_queries, self.process_latencies)
+        print("Finished constructing SUT.")
+        self.qsl = get_kits_QSL(preprocessed_data_dir, performance_count)
+
+    def to_tensor(self, my_array):
+        """
+        Transforms my_array into tensor form backend understands
+        Implementation may differ as backend's need
+        """
+        return my_array
+
+    def from_tensor(self, my_tensor):
+        """
+        Transforms my_tensor backend worked on into numpy friendly array
+        Implementation may differ as backend's need
+        """
+        return my_tensor
+
+    def do_infer(self, input_tensor):
+        """
+        Performs backend specific inference upon input_tensor
+        Need specific implementation for the backend
+        """
+        assert False, "Please implement!"
+
+    @infu.runtime_measure
+    def infer_single_query(self, query, mystr):
```

Comment:
Slicing is done after query is received from loadgen.  Does it imply it's legit to reuse convolution output of overlapped area between dices, since it comes from a single query?

Reply:
I think it's more of a question to Working Group. Outside the MLPerf, doing inference with this trained network upon these input tensors can be implemented in many different arbitrary ways. In the MLPerf though, I think all the stakeholders should agree on what technique is used. :) So it's good idea to bring any specific technique in mind to WG and get the agreement.